extension/source/Controllers/Network/NetworkController.ts

```diff
@@ -0,0 +1,249 @@
+    return this.state.chainId;
+  }
+
+  /**
+   * Called by orchestrator once while initializing the class
+   * @param providerHandlers - JRPC handlers for provider
+   * @returns - provider - Returns the providerProxy
+   */
+  public initializeProvider(
+    providerHandlers: IProviderHandlers,
+  ): SafeEventEmitterProvider {
+    this._baseProviderHandlers = providerHandlers;
+    this.configureProvider();
+    this.lookupNetwork(); // Not awaiting this, because we don't want to block the initialization
+    return this._providerProxy;
+  }
+
+  getProvider(): SafeEventEmitterProvider {
+    return this._providerProxy;
+  }
+
+  setProviderConfig(config: ProviderConfig): void {
+    this.update({
+      providerConfig: { ...config },
+    });
+    this.refreshNetwork();
+  }
+
+  getProviderConfig(): ProviderConfig {
+    return this.state.providerConfig;
+  }
+
+  /**
+   * Refreshes the current network code
+   */
+  async lookupNetwork(): Promise<void> {
+    const { rpcTarget, chainId } = this.getProviderConfig();
+    if (!chainId || !rpcTarget || !this._provider) {
+      this.update({ chainId: 'loading', properties: {} });
+      return;
+    }
+    const query = this.ethQuery;
+    if (!query) {
+      return;
+    }
+    const releaseLock = await this.mutex.acquire();
+    return new Promise((resolve, reject) => {
+      // info_get_status
+      query.sendAsync(
+        { method: 'net_version' },
+        (error: Error, network: unknown) => {
+          releaseLock();
+          if (error) {
+            this.update({
+              chainId: 'loading',
+              properties: {
+                EIPS: { 1559: undefined },
+              },
+            });
+            reject(error);
+          }
+
+          this.update({
+            // Network is returned as a string (base 10)
+            chainId: `0x${Number.parseInt(network as string, 16).toString()}`,
+          });
+          // Don't need to wait for this
+          this.getEIP1559Compatibility();
+          this.emit('networkDidChange');
+          resolve();
+        },
+      );
+    });
+  }
+
+  async getEIP1559Compatibility(): Promise<boolean> {
+    const { EIPS } = this.state.properties;
+    // log.info('checking eip 1559 compatibility', EIPS[1559])
+    if ((EIPS as any)[1559] !== undefined) {
+      return (EIPS as any)[1559];
+    }
+    return new Promise((resolve, reject) => {
+      this.ethQuery.sendAsync(
+        {
+          method: 'eth_getBlockByNumber',
+          params: ['latest', false],
+        },
+        (err: unknown, latestBlock: unknown) => {
+          if (err) reject(err);
+          const supportsEIP1559 =
+            (latestBlock as Block) &&
+            (latestBlock as Block).baseFeePerGas !== undefined;
+          this.update({
+            properties: {
+              EIPS: { 1559: supportsEIP1559 },
+            },
+          });
+          resolve(supportsEIP1559);
+        },
+      );
+    });
```

Comment:
Haven't been commenting on all these but there are a lot of examples of code like this:

```ts
return new Promise((resolve, reject) => {
  someAsyncFunction((err, value) => {
    if (err) return reject(err);
    // do some processing on `value`
    resolve();
  });
});
```

Which should be like this:

```ts
const value = await new Promise((resolve, reject) => {
  someAsyncFunction((err, value) => {
    if (err) { reject(err); }
    else { resolve(value); }
  });
});

// do some processing on `value`
```

Also can be further improved by promisifying `someAsyncFunction` so you can simply write:

```ts
const value = await promisifedFunction();
// do some processing on `value`
```

Reply:
Added promise based augmentation method `request`